core/src/cluster_info.rs

```diff
@@ -1200,22 +1202,24 @@ impl ClusterInfo {
     /// all tvu peers with valid gossip addrs that likely have the slot being requested
     pub fn repair_peers(&self, slot: Slot) -> Vec<ContactInfo> {
         let mut time = Measure::start("repair_peers");
-        let ret = ClusterInfo::tvu_peers(self)
-            .into_iter()
-            .filter(|x| {
-                x.id != self.id()
-                    && x.shred_version == self.my_shred_version()
-                    && ContactInfo::is_valid_address(&x.serve_repair)
-                    && {
-                        self.get_lowest_slot_for_node(&x.id, None, |lowest_slot, _| {
-                            lowest_slot.lowest <= slot
-                        })
-                        .unwrap_or_else(|| /* fallback to legacy behavior */ true)
-                    }
-            })
-            .collect();
+        // self.tvu_peers() already filters on:
+        //   node.id != self.id() &&
+        //     node.shred_verion == self.my_shred_version()
+        let nodes = {
+            let gossip = self.gossip.read().unwrap();
+            self.tvu_peers()
```

Comment:
@behzadnouri this can cause a deadlock on machines with write biased rwlocks if a writer gets queued for the rwlock before the read lock is acquired inside `self.tvu_peers()`.

Reply:
hehe good catch! glad that local cluster debugging also reaped other benefits 🎉

Reply:
Thanks for catching this. Sent out https://github.com/solana-labs/solana/pull/13973